contracts/SolverTrampoline.sol

```diff
@@ -17,17 +23,41 @@ contract SolverTrampoline {
     /// @dev The domain separator for signing EIP-712 settlements.
     bytes32 public immutable domainSeparator;
 
+    /// @dev Nonce by solver address.
+    mapping(address => uint256) public nonces;
+
     constructor(Settlement settlementContract_) {
         settlementContract = settlementContract_;
         solverAuthenticator = settlementContract_.authenticator();
-        
+
         domainSeparator = keccak256(abi.encode(
             keccak256("EIP712Domain(uint256 chainId,address verifyingContract)"),
             block.chainid,
             address(this)
         ));
     }
 
+    /// @dev Executes a settlement on behalf of a solver.
+    function settle(bytes calldata settlement, uint256 nonce, bytes32 r, bytes32 s, uint8 v) external {
+        bytes32 messageDigest = settlementMessage(settlement, nonce);
+        address solver = ecrecover(messageDigest, v, r, s);
+        if (solver == address(0) || !solverAuthenticator.isSolver(solver)) {
+            revert Unauthorized();
+        }
+        if (nonce != nonces[solver]) {
```

Comment:
I wonder if it is limiting to expect nonce to be strictly sequential (due to stuck txs or the like). What would be the trade-off to having a bitmap where ordering of nonces wouldn't matter (just that they are not reused).

Reply:
> I wonder if it is limiting to expect nonce to be strictly sequential (due to stuck txs or the like).

What I envisioned would be that solvers could cancel their nonces.

> What would be the trade-off to having a bitmap where ordering of nonces wouldn't matter

Code complexity.

Reply:
`if nonce[solver][nonce] != 0` doesn't seem much more complex to me, but it would require an extra cold sstore on every run, which I accept as a good enough reason.

Reply:
Ah, when you mentioned bitmap, I understood it as packing multiple nonces into a single slot in order to reduce gas costs. Something like:
```
uint256 index = nonce >> 5;
uint256 mask = 1 << (nonce & 0x1f);
uint256 nonceBitmap = nonce[solver][index];
require(nonceBitmap & mask == 0);
nonce[solver][index] = nonceBitmap | mask;
```

This would only cost an addition zero->nonzero `SSTORE` once per 32 settlements.

Reply:
Admittedly... it is not thaaaat much additional complexity.

Reply:
Something that wouldn't cost too much more could be letting the solver manage the nonce.
`settle` could take `current_nonce` and `next_nonce`. The transaction can only be executed if the current nonce matches. This has a smaller gas impact than an extra sstore and allows for some more complicated logic. For example allowing a transaction to be executed only if a previous one was executed. To execute two independent transactions A and B simultaneously from current nonce `n`, we would send four signatures to the executor: `sign(A, n, n+1)`, `sign(B, n+1, n+2)`, and the two transactions to skip a failing transaction: `sign(A, n, n+2)`, `sign(B, n, n+2)`. (If we don't trust the executor it can still skip a valid transaction, but we'd just lose the time needed to notice that transaction was executed and signing another for `n+2 -> n+3`.
In this example we should probably still make sure that `current_nonce < next_nonce` to avoid that old transactions can be reused.

Reply:
We talked about this in the SC meeting and decided:
- We should revert when the transaction reverts here:
  - Relayers sometimes don't charge for reverted transactions, which is nice for us
  - If the settlement becomes invalid, then valid again within the deadline, we want it to execute.
- Sequential nonce is fine, since we only have one transaction per solver at a time anyway.
- We will add a "deadline" to the transaction, so no manual cancellation would be needed.
- We will add a utility `cancel` function that increments a solver nonce.